### DISCUSSION TITLE
Oversight/incorrect documentation? Config file suggests a custom key provider can be used but there is nothing in the docs nor does the code indicate this is possible.

### DISCUSSION THREAD

**Discussion by stevenmaguire** [spatie/laravel-ciphersweet#28] [Q&A] [answered]
Is this an oversight? Is it possible to build your own key provider or not?

![CleanShot 2023-03-29 at 17 24 15@2x](https://user-images.githubusercontent.com/1851973/228681692-a6fe4f59-d2a6-4643-922a-4e9a11a8e956.png)
![CleanShot 2023-03-29 at 17 26 24@2x](https://user-images.githubusercontent.com/1851973/228681703-44ecfe6f-c5c6-49d4-939c-bbc8fe5c256a.png)

**Comment by freekmurze** [comment]
@riasvdv could you take a look at this one?

**Comment by riasvdv** [comment, marked as answer]
@stevenmaguire This was indeed missing. I've now added it to the package: https://github.com/spatie/laravel-ciphersweet#implementing-a-custom-key-provider

**Reply by stevenmaguire** [reply]
Thank you for the confirmation and prompt response!

**Comment by stevenmaguire** [comment]
As a follow-up - and I am happy to open a new discussion if that is preferred - do you think similar support for custom backends is reasonable? 

I'm specifically thinking about the MultiTenant support from the underlying package. It requires a KeyProvider and a Backend that implement specific MultiTenant interfaces. The current stock options do not appear to satisfy that. If I can provide a custom KeyProvider and custom Backend it seems that functionality will be accessible.

**Reply by riasvdv** [reply]
We don’t need this functionality ourselves, but I’d gladly merge a PR with tests that adds this
